readme.md
# This project is a fork of initial project called CardDavMATE hosting on http://www.inf-it.com/



changelog.txt
CardDavMATE Changelog

version 0.9.6 [2012-04-25]:
- note: if you use Davical with cross-domain setup see the modified Apache configuration in misc/config_davical.txt (solved problem with CardDAV-Sync from Marten Gajda)
- note: if you use MacOS X Lion calendarserver it is recommended to re-patch your installation (see misc/readme_lion.txt and the changelog below)
- configuration - changed default syncInterval from 15000 to 30000
- added support for automatic detection of crossDomain option in globalAccountSettings, globalNetworkCheckSettings and globalNetworkAccountSettings (if not set to boolean true/false it is automatically detected; default is null - autodetect)
- added IE9+ support (non cross-domain setup only /IE9 & jQuery limitation/; minor graphics glitches due to missing CSS support for "disabled" html elements)
- added support for prefix and suffix (vCard N attribute)
- added Czech localization (cs_CZ)
- added Danish localization (da_DK) - thanks Niels Bo Andersen
- fixed search with upper case letters not present in globalSearchTransformAlphabet
- fixed parsing of double quoted vCard param values (who use this?)
- fixed Andorra address fields
- fixed minor graphics glitch with update notification after logout and new login
- updated auth module to support multiple resources in response + updated auth/config.inc with modified templates
- updated misc/config_davical.txt (better handling of preflighted OPTION requests)
- updated misc/calendarserver_CardDavMATE.diff (support for cross-domain queries with X-Requested-With header /upcoming Safari 5.2/)
- updated default globalSearchTransformAlphabet in config.js
- updated CSS detection rule for iPad (new screen resolution for iPad 3)
- updated jQuery to 1.7.2
- updated jQuery-UI to 1.8.19
- replaced .attr('value'[, ...]) with .val([...])
- many other fixes

version 0.9.5 [2012-03-12]:
- note: this release contains new configuration option (always use the latest config.js)
- configuration - added globalSearchTransformAlphabet setting - used for transformation of non-ASCII characters to ASCII (for search support)
- added search support (search is performed on the addressbook values in the interface + category names)
- added language selector to login screen (use globalInterfaceLanguage to set the preselected language)
- added hungarian localization (hu_HU)
- scrollable resource list (only if needed)
- minor bug fixes

version 0.9.4 [2012-02-29]:
- fixed vCard parameter parsing (TYPE=X,Y now works correctly again)
- fixed date parsing for BDAY and X-ANNIVERSARY attributes (dates with missing '-' caused non-RFC vCard error)
- correct "\n" unescaping in vCard values if the previous character is '\'
- updated phone parameter definitions (iOS5 uses "TYPE=voice" in addition of other types)
- correct header removing from addressbook if contact is removed
- changed default value for undefined phone type from "main" to "cell" ("main" is Apple specific)
- added "TYPE=internet" into EMAIL parameters if it is missing (better non-RFC vCard compatibility)
- added "fax" phone type (in addition of "home fax" and "work fax")
- use of .prop() instead of .attr() to set 'readonly', 'disabled', 'selected', 'checked' and 'src' properties
- minor fixes for Slovak and German (thanks Thomas Scheel) localizations
- other minor fixes

version 0.9.3 [2012-02-17]:
- added german localization (de_DE) - thanks Marten Gajda
- fixed invalid switch to PROPFIND if REPORT/sync-collection is unsupported (+ related bugfixes)
- changed DOM operations to fix Firefox only bugs with 0.9.2 (Firefox or jQuery bug?)
- fixed "active editor -> logout -> login" problem (inactive resource/addressbook list)

version 0.9.2 [2012-02-13]:
- security fix: correct globalAccountSettings cleanup after logout
- note: this release contains new configuration options (always use the latest config.js)
- configuration - increased default timeout value for configuration account templates from 10 to 30 seconds
- configuration - changed default configuration for Davical - works also if Davical is installed into subdirectory (instead of domain root directory)
- configuration - added globalContactStoreFN option - support for custom FN formatting instead of fixed 'prefix,last,middle,first,suffix' (see config.js)
- configuration - added globalCompatibility option - customizations for 3rd party clients
- configuration - globalCompatibility: added anniversaryOutputFormat option - supported values: 'apple' (default) and 'other' (see config.js)
- configuration - added globalInterfaceLanguage option (see config.js and the changelog below)
- added localization support (new localizations are welcome!)
- added english localization (en_US)
- added slovak localization (sk_SK)
- added italian localization (it_IT) - thanks Luca Ferrario
- added support for Oracle, SabreDav and probably many other CardDav servers
- added support for CATEGORIES attribute with auto-suggest for categories in active collection (category separator: Enter)
- 100% performance improvement for PUT requests (add/edit/move/... operations) - get the modified Etag from PUT response header instead of additional REPORT requests (if the server support this feature, otherwise get the modified vCard with new Etag from the server)
- notable performance improvements for DOM operations (contact loading and manipulation)
- correct '\' character escaping in vCard values
- removed write support for X-EVOLUTION-* attributes (caused many problems and Evolution already support the X-* attributes without "-EVOLUTION" prefix)
- many other fixes

version 0.9.1 [2012-01-29]:

interface.js
/*
CardDavMATE - CardDav Web Client
Copyright (C) 2011-2012 Jan Mate <jan.mate@inf-it.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

function processEditorElements(processingType, inputIsReadonly)
{
	var cssShowAsTxtClass='element_show_as_text';
	var cssGrayedTxt='element_grayed';
	var cssElementNoDisplay='element_no_display';
	var cssElementHide='element_hide';

	if(processingType=='hide')
	{
		$('[id=vcard_editor]').attr('data-editor-state','show');
		disabled=true;
		readonly=true;
	}
	else
	{
		$('[id=vcard_editor]').attr('data-editor-state','edit');
		disabled=false;
		readonly=false;
	}

	// company checkbox and text
	var tmp=$('[id="vcard_editor"]').find('[data-type="isorg"]');
	tmp.prop('disabled',disabled);
	if(processingType=='hide' && !tmp.is(':checked'))
		tmp.parent().addClass(cssGrayedTxt);
	else
		tmp.parent().removeClass(cssGrayedTxt);

	$('[data-type^="date_"]').prop('disabled', disabled || readonly);

	// family name, given name, and organization name
	var typeList = new Array('family','given','middle','nickname','date_bday','date_anniversary','tags','title','department','org');
	for(i=0;i<typeList.length;i++)
		$('[id=vcard_editor]').find('[data-type="'+typeList[i]+'"]').prop('readonly',readonly);

	var tmp=$('[id=vcard_editor]').find('#tags_tag');
	tmp.prop('readonly',readonly);
	if(readonly)
		tmp.closest('div.tagsinput').addClass('readonly');
	else
		tmp.closest('div.tagsinput').removeClass('readonly');

	// set the visibility of the buttons
	var tmp=$('[id=vcard_editor]');
	if(processingType=='hide')
	{
		if(inputIsReadonly!=true)
			tmp.find('[data-type="edit"]').removeClass(cssElementNoDisplay);
		else
			tmp.find('[data-type="edit"]').addClass(cssElementNoDisplay);

		tmp.find('[data-type="save"]').addClass(cssElementNoDisplay);
		tmp.find('[data-type="cancel"]').addClass(cssElementNoDisplay);
		tmp.find('[data-type="delete_from_group"]').addClass(cssElementNoDisplay);
		tmp.find('[data-type="delete"]').addClass(cssElementNoDisplay);
	}
	else if(processingType=='add')
	{
		tmp.find('[data-type="edit"]').addClass(cssElementNoDisplay);
		tmp.find('[data-type="cancel"]').removeClass(cssElementNoDisplay);
		tmp.find('[data-type="delete_from_group"]').addClass(cssElementNoDisplay);
		tmp.find('[data-type="delete"]').addClass(cssElementNoDisplay);
	}
	else
	{
		tmp.find('[data-type="edit"]').addClass(cssElementNoDisplay);
		tmp.find('[data-type="save"]').removeClass(cssElementNoDisplay);
		tmp.find('[data-type="cancel"]').removeClass(cssElementNoDisplay);
		// show "Delete from Group" only if there is an active contact group
		if(globalResourceList.getLoadedAddressbook().filterUID!='')
			tmp.find('[data-type="delete_from_group"]').removeClass(cssElementNoDisplay);
		tmp.find('[data-type="delete"]').removeClass(cssElementNoDisplay);
	}


	var typeList = new Array('\\%address','\\%phone','\\%email','\\%url','\\%person','\\%im','\\%categories','\\%note');
	for(i=0;i<typeList.length;i++)
	{
		found_non_empty=0;

		tmp=$('[id=vcard_editor]').find('[data-type="'+typeList[i]+'"]');
		tmp.each(
			function(index,element)
			{
				var tmp=$(element).find('[data-type="value"]');

				var found=0;
				// check if there is any data present (if not, whe hide the element)
				if($(element).attr('data-type')=='%address')	// address is handled specially
					tmp.each(
						function(index,element)
						{
							if($(element).attr('data-addr-field')!='' && $(element).attr('data-addr-field')!='country' && $(element).val()!='')
							{
								found=1;
								return false;
							}
						}
					);
				else if(tmp.attr('value')!='')	// other elements (not address)
					found=1;


				if(processingType=='hide')
				{
					if(!found)
					{
						$(element).addClass(cssElementNoDisplay);
					}
					else
					{
						$(element).find('[data-type="\\%add"]').find('input[type="image"]').addClass(cssElementNoDisplay);
						$(element).find('[data-type="\\%del"]').find('input[type="image"]').addClass(cssElementNoDisplay);
						$(element).find('select').prop('disabled',disabled);
						$(element).find('textarea').prop('disabled',disabled);
						tmp.prop('readonly',readonly);
						found_non_empty=1;
					}
					
				}	
				else	// 'show'
				{
					$(element).removeClass(cssElementNoDisplay);
					$(element).find('[data-type="\\%add"]').find('input[type="image"]').removeClass(cssElementNoDisplay);
					$(element).find('[data-type="\\%del"]').find('input[type="image"]').removeClass(cssElementNoDisplay);
					$(element).find('select').prop('disabled',disabled);
					$(element).find('textarea').prop('disabled',disabled);
					tmp.prop('readonly',readonly);
				}
			}
		);

		if(!found_non_empty)
		{
			if(processingType=='hide')
				tmp.prev().addClass(cssElementNoDisplay);
			else
				tmp.prev().removeClass(cssElementNoDisplay);
		}
	}
}

function editor_cleanup(inputLoadEmpty)
{
	cleanupRegexEnvironment();

	// Cleanup the editor and select the default country
	$('#ABContact').html(cleanVcardTemplate);

	var tmp=$('[id="vcard_editor"] [data-type="\\%address"]');
	var tmp_select=tmp.find('[data-type="\\%country"]').attr('data-autoselect');
	if(tmp_select!='')
	{
		tmp.find('[data-type="\\%country"]').children('[data-type="'+jqueryEscapeSelector(tmp_select)+'"]').prop('selected',true);
		tmp.find('[data-autoselect]').change();
	}

	$('#tags').tagsInput({
		'height': null,
		'width': '448px',
		'placeholderColor': '#e0e0e0',
		'useNativePlaceholder': true,
		'defaultText': localization[globalInterfaceLanguage].addCategory,
		'delimiter': ',',
		'allowDelimiterInValue': true,	// if true delimiter is escaped with '\' ('\' is escaped as '\\')
		'trimInput': false,
		'autocomplete_url': globalAddressbookList.getABCategories(),
		'autocomplete': {
			'autoFocus': true,
			'minLength': 0
		},
		'onAddTag': function(value)
		{
			// copy the array
			var xList=globalAddressbookList.getABCategories();
			var currentTags=$(this).val().splitCustom(',');
			for(var i=xList.length-1; i>=0; i--)
			{
				for(var j=0; j<currentTags.length; j++)
					if(xList[i] == currentTags[j])
						xList.splice(i, 1);
			}
			$('#tags_tag').autocomplete('option', 'source', xList);
		},
		'onRemoveTag': function()
		{
			// copy the array
			var xList=globalAddressbookList.getABCategories();
			var currentTags=$(this).val().splitCustom(',');
			for(var i=xList.length-1; i>=0; i--)
			{
				for(var j=0; j<currentTags.length; j++)
					if(xList[i] == currentTags[j])
						xList.splice(i, 1);
			}
			$('#tags_tag').autocomplete('option', 'source', xList);
		}
	});

	if(inputLoadEmpty==true)
		$('#EditorBox').fadeTo(100,1);
}

function animate_message(messageSelector,messageTextSelector,duration,operation)
{
	if(operation==undefined)
		operation='+=';
	var height=$(messageTextSelector).height()+14;
	var animation=500;

	$(messageSelector).animate({'max-height': height+'px', height: (operation==undefined ? '+=' : operation)+height+'px'}, animation,
		function()
		{
			if(operation=='+=')
				setTimeout(function(){animate_message(messageSelector,messageTextSelector,0,'-=');},duration);
		}
	);
	return duration+2*animation;
}

function show_editor_message(inputPosition,inputSetClass,inputMessage,inputDuration)
{
	if(inputPosition==undefined || inputPosition=='in')
	{
		messageSelector='#ABInMessage';
		messageTextSelector='#ABInMessageText';
	}
	else
	{
		messageSelector='#ABMessage';
		messageTextSelector='#ABMessageText';
	}

	$(messageTextSelector).attr('class',inputSetClass);
	$(messageTextSelector).text(inputMessage);
	return animate_message(messageSelector,messageTextSelector,inputDuration);
}

function set_address_country(inputSelectedAddressObj)
{
	var selectedCountry=$(inputSelectedAddressObj).find('option:selected').attr('data-type');
	var addressElement=$(inputSelectedAddressObj).closest('[data-type="\\%address"]');

	// cleanup the data-addr-fields and placeholders
	addressElement.find('[data-addr-fid]').each(
		function(index,element)
		{
			$(element).find('input').attr('data-addr-field','');
			$(element).find('input').attr('placeholder','');
		}
	);

	if(addressTypes[selectedCountry]!=undefined)
	{
		for(var i=0;i<addressTypes[selectedCountry].length;i++)
		{
			if(addressTypes[selectedCountry][i]['type']=='input')
			{
				var tmp=addressElement.find('[data-addr-fid="'+jqueryEscapeSelector(addressTypes[selectedCountry][i]['fid'])+'"]').find('input');
				tmp.attr('data-addr-field',addressTypes[selectedCountry][i]['data-addr-field']);
				tmp.attr('placeholder',addressTypes[selectedCountry][i]['placeholder']);
			}
			else if(addressTypes[selectedCountry][i]['type']=='country')
			{
				var tmp=addressElement.find('[data-type="\\%country"]');
				tmp.find('option[data-type]').prop('selected',false);
				tmp.find('option[data-type="'+jqueryEscapeSelector(selectedCountry)+'"]').prop('selected',true);

				// the country selector is in wrong container -> we need to move it
				if(addressTypes[selectedCountry][i]['fid']!=tmp.closest('[data-addr-fid]').attr('data-addr-fid'))
					$(addressElement).find('[data-addr-fid="'+jqueryEscapeSelector(addressTypes[selectedCountry][i]['fid'])+'"]').append(tmp); 
			}
		}
	}

	// hide the unused fields by changing the CSS
	addressElement.find('[data-type="container"]').each(
		function(index,element)
		{
			var found=0;
			$(element).find('[data-addr-field]').each(
				function(index,element)
				{
					if($(element).attr('data-addr-field')!='')
					{
						found=1;
						return false;
					}
				}
			);

			if(found)
				$(element).removeClass('element_no_display_af');
			else
				$(element).addClass('element_no_display_af');
		}
	);

}

function add_element(inputElementID, inputParentSelector, newElementSelector, inputAddClassSelector, inputDelClassSelector, maxElements, newElementID)
{
	// allow only maxElements items for this attribute 
	if((count=inputElementID.closest(inputParentSelector).parent().children(inputParentSelector).length) < maxElements)
	{
		newElement=$(newElementSelector).clone().wrap('<div>');

		// unselect each selected element
		newElement.find('option').prop('selected',false);
		// remove the form values
		newElement.find('[data-type="value"]').attr('value','');
		// add the data-id value
		newElement.find(inputParentSelector).attr("data-id",newElementID);
		// disable the "add" on last element if maximum count is reached
		if(count==maxElements-1)
			newElement.find(inputAddClassSelector).css('visibility','hidden');

		newElement=newElement.closest(inputParentSelector).parent().html();

		// disable the "add" button on the current element
		inputElementID.filter(inputAddClassSelector).css('visibility','hidden');
		// add the new element (with enabled "add" button)
		inputElementID.parent().after(newElement);
		// enable the "del" button on all elements
		$(inputParentSelector).find(inputDelClassSelector).css('visibility','');

		if(inputParentSelector=='[data-type="\\%address"]')
		{
			// execute the "autoselect"
			var tmp=inputElementID.closest(inputParentSelector).next();
			var tmp_select=tmp.find('[data-autoselect]').attr('data-autoselect');
			if(tmp_select!=null)
			{
				tmp.find('[data-type="\\%country"]').children('[data-type="'+jqueryEscapeSelector(tmp_select)+'"]').prop('selected',true);
				tmp.find('[data-autoselect]').change();
			}
		}
		return true;
	}
	else
		return false;
}

function del_element(inputElementID, inputParentSelector, inputAddClassSelector, inputDelClassSelector)
{
	// all elements except the last can be removed
	if(inputElementID.closest(inputParentSelector).siblings(inputParentSelector).length > 0)
	{
		inputElementID.closest(inputParentSelector).remove();
		// enable the "add" button on last element
		$(inputParentSelector).last().find(inputAddClassSelector).css('visibility','');
		// disable the "del" button if only one element is present
		if($(inputParentSelector).length==1)
			$(inputParentSelector).last().find(inputDelClassSelector).css('visibility','hidden');
	}
}

var globalCounter = new Object();
globalCounter['phoneID']=1;
globalCounter['emailID']=1;
globalCounter['urlID']=1;
globalCounter['personID']=1;
globalCounter['imID']=1;
globalCounter['addressID']=1;

$(document).on('keyup change', '[data-type^="date_"]', function(){
	if(!$(this).prop('readonly') && !$(this).prop('disabled'))
	{
		var valid=true;

		if($(this).val()!='')
		{
			try {$.datepicker.parseDate(globalDatepickerFormat, $(this).val())}
			catch (e) {valid=false}
		}

		if(valid)
			$(this).parent().find('img').css('display','none');
		else
			$(this).parent().find('img').css('display','inline');
	}
});

$(document).on('focus', '[data-type^="date_"]', function(){if(!$(this).hasClass('hasDatepicker')){$(this).datepicker({disabled: $(this).prop('readonly') || $(this).prop('disabled'), showMonthAfterYear: true, prevText: '', nextText: '', monthNamesShort: ['01','02','03','04','05','06','07','08','09','10','11','12'], dateFormat: globalDatepickerFormat, defaultDate: '-'+Math.round(30*365.25-1), minDate: '-120y', maxDate: '+0', yearRange: 'c-120:+0', firstDay: 0, changeMonth: true, changeYear: true,
	beforeShow: function(input, inst)	// set the datepicker value if the date is out of range (min/max)
	{
		var valid=true;
		try {var currentDate=$.datepicker.parseDate(globalDatepickerFormat, $(this).val())}
		catch (e) {valid=false}

		if(valid==true)
		{
			var minDateText=$(this).datepicker('option', 'dateFormat', globalDatepickerFormat).datepicker('option', 'minDate');
			var maxDateText=$(this).datepicker('option', 'dateFormat', globalDatepickerFormat).datepicker('option', 'maxDate');

			var minDate=$.datepicker.parseDate(globalDatepickerFormat, minDateText);
			var maxDate=$.datepicker.parseDate(globalDatepickerFormat, maxDateText);

			if(currentDate<minDate)
				$(this).val(minDateText);
			else if(currentDate>maxDate)
				$(this).val(maxDateText);
		}
	},
	onClose: function(dateText, inst)	// set the datepicker value if the date is out of range (min/max) and reset the value to proper format (for example 'yy-mm-dd' allows '2000-1-1' -> we need to reset the value to '2000-01-01')
	{
		var valid=true;
		try {var currentDate=$.datepicker.parseDate(globalDatepickerFormat, dateText)}
		catch (e) {valid=false}

		if(valid==true)
		{
			var minDateText=$(this).datepicker('option', 'dateFormat', globalDatepickerFormat).datepicker('option', 'minDate');
			var maxDateText=$(this).datepicker('option', 'dateFormat', globalDatepickerFormat).datepicker('option', 'maxDate');

			var minDate=$.datepicker.parseDate(globalDatepickerFormat, minDateText);
			var maxDate=$.datepicker.parseDate(globalDatepickerFormat, maxDateText);

			if(currentDate<minDate)
				$(this).val(minDateText);
			else if(currentDate>maxDate)
				$(this).val(maxDateText);
			else
				$(this).val($.datepicker.formatDate(globalDatepickerFormat, currentDate));
		}
		$(this).focus();
	}
});
$(this).mousedown(function(){
	if($(this).datepicker('widget').is(':visible'))
		$(this).datepicker('hide');
	else
		$(this).datepicker('show');
})}});

phoneMax=20;
$(document).on('click', '[data-type="\\%phone"] [data-type="\\%add"] input', function(ignoreMaxElements){add_element($(this).parent(),'[data-type="\\%phone"]','[data-type="\\%phone"]:last','[data-type="\\%add"]','[data-type="\\%del"]',phoneMax,globalCounter['phoneID']++)});
$(document).on('click', '[data-type="\\%phone"] [data-type="\\%del"] input', function(){del_element($(this).parent(),'[data-type="\\%phone"]','[data-type="\\%add"]','[data-type="\\%del"]')});
//$('[data-type="\\%phone"]').children().filter('[data-type="\\%add"]').click();

emailMax=20;
$(document).on('click', '[data-type="\\%email"] [data-type="\\%add"] input', function(ignoreMaxElements){add_element($(this).parent(),'[data-type="\\%email"]','[data-type="\\%email"]:last','[data-type="\\%add"]','[data-type="\\%del"]',emailMax,globalCounter['emailID']++)});
$(document).on('click', '[data-type="\\%email"] [data-type="\\%del"] input', function(){del_element($(this).parent(),'[data-type="\\%email"]','[data-type="\\%add"]','[data-type="\\%del"]')});
//$('[data-type="\\%email"]').children().filter('[data-type="\\%add"]').click();

urlMax=20;
$(document).on('click', '[data-type="\\%url"] [data-type="\\%add"] input', function(ignoreMaxElements){add_element($(this).parent(),'[data-type="\\%url"]','[data-type="\\%url"]:last','[data-type="\\%add"]','[data-type="\\%del"]',urlMax,globalCounter['urlID']++)});
$(document).on('click', '[data-type="\\%url"] [data-type="\\%del"] input', function(){del_element($(this).parent(),'[data-type="\\%url"]','[data-type="\\%add"]','[data-type="\\%del"]')});
//$('[data-type="\\%url"]').children().filter('[data-type="\\%add"]').click();

personMax=20;
$(document).on('click', '[data-type="\\%person"] [data-type="\\%add"] input', function(ignoreMaxElements){add_element($(this).parent(),'[data-type="\\%person"]','[data-type="\\%person"]:last','[data-type="\\%add"]','[data-type="\\%del"]',personMax,globalCounter['personID']++)});
$(document).on('click', '[data-type="\\%person"] [data-type="\\%del"] input', function(){del_element($(this).parent(),'[data-type="\\%person"]','[data-type="\\%add"]','[data-type="\\%del"]')});
//$('[data-type="\\%person"]').children().filter('[data-type="\\%add"]').click();

imMax=20;
$(document).on('click', '[data-type="\\%im"] [data-type="\\%add"] input', function(ignoreMaxElements){add_element($(this).parent(),'[data-type="\\%im"]','[data-type="\\%im"]:last','[data-type="\\%add"]','[data-type="\\%del"]',imMax,globalCounter['imID']++)});
$(document).on('click', '[data-type="\\%im"] [data-type="\\%del"] input', function(){del_element($(this).parent(),'[data-type="\\%im"]','[data-type="\\%add"]','[data-type="\\%del"]')});
//$('[data-type="\\%im"]').children().filter('[data-type="\\%add"]').click();

addrMax=20;
$(document).on('click', '[data-type="\\%address"] [data-type="\\%add"] input', function(ignoreMaxElements){add_element($(this).parent(),'[data-type="\\%address"]','[data-type="\\%address"]:last','[data-type="\\%add"]','[data-type="\\%del"]',imMax,globalCounter['addressID']++)});
$(document).on('click', '[data-type="\\%address"] [data-type="\\%del"] input', function(){del_element($(this).parent(),'[data-type="\\%address"]','[data-type="\\%add"]','[data-type="\\%del"]')});
//$('[data-type="\\%address"]').children().filter('[data-type="\\%add"]').click();
$(document).on('change', '[data-type="\\%address"] [data-type="\\%country"]', function(){set_address_country(this)});
